src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -11848,6 +11904,99 @@ class CL1<T>
                 );
         }
 
+        [Fact]
+        public void Lambda_18()
+        {
+            var source =
+@"delegate T D<T>(T t) where T : class;
+class C
+{
+    static void F()
+    {
+        var d1 = (D<string?>)((string s1) =>
+            {
+                s1 = null;
+                return s1;
+            });
+        var d2 = (D<string>)((string? s2) =>
+            {
+                s2.ToString();
+                return s2;
+            });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,22): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //                 s1 = null;
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(8, 22),
+                // (13,17): warning CS8602: Possible dereference of a null reference.
+                //                 s2.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(13, 17),
+                // (14,24): warning CS8603: Possible null reference return.
+                //                 return s2;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "s2").WithLocation(14, 24));
+        }
+
+        /// <summary>
+        /// Inferred nullability of captured variables outside lambda
+        /// should be independent of inferred nullability inside lambda.
+        /// </summary>
+        [Fact]
+        public void Lambda_19()
+        {
+            var source =
+@"using System;
+class C
+{
+    static void F1(object? x1, object y1)
+    {
+        object z1 = y1;
+        Action f = () =>
+        {
+            z1 = x1; // warning
+        };
+        f();
+        z1.ToString();
+    }
+    static void F2(object? x2, object y2)
+    {
+        object z2 = x2; // warning
+        Action f = () =>
+        {
+            z2 = y2;
+        };
+        f();
+        z2.ToString(); // warning
+    }
+    static void F3(object? x3, object y3)
+    {
+        object z3 = y3;
+        if (x3 == null) return;
+        Action f = () =>
+        {
+            z3 = x3; // warning
+        };
+        f();
+        z3.ToString();
```

Comment:
>z3 [](start = 8, length = 2)

Diagnostic here too. Note that this would be a non-W warning. It would be a shame if there were no non-W warnings in this code to catch the possible null dereference.